File: sdks/java/src/main/java/dev/atap/sdk/ATAPClient.java
package dev.atap.sdk;

import dev.atap.sdk.api.*;
import dev.atap.sdk.crypto.Ed25519KeyPair;
import dev.atap.sdk.http.ATAPHttpClient;
import dev.atap.sdk.oauth.TokenManager;

import java.time.Duration;
import java.util.List;
import java.util.Map;

/**
 * High-level client for the ATAP platform.
 *
 * <p>Usage for agent/machine (client_credentials):
 * <pre>{@code
 * ATAPClient client = ATAPClient.builder()
 *     .baseUrl("http://localhost:8080")
 *     .did("did:web:localhost%3A8080:agent:abc")
 *     .privateKey("<base64 Ed25519 seed>")
 *     .clientSecret("atap_...")
 *     .build();
 * }</pre>
 */
public class ATAPClient implements AutoCloseable {

    private final ATAPHttpClient httpClient;
    private final String did;
    private final String platformDomain;
    private final Ed25519KeyPair keyPair;
    private final TokenManager tokenManager;

    /**
     * API modules.
     */
    public final EntityApi entities;
    public final ApprovalApi approvals;
    public final RevocationApi revocations;
    public final DIDCommApi didcomm;
    public final CredentialApi credentials;
    public final DiscoveryApi discovery;

    ATAPClient(String baseUrl, String did, Ed25519KeyPair keyPair,
               String clientSecret, List<String> scopes,
               String platformDomain, Duration timeout) {
        this.httpClient = new ATAPHttpClient(baseUrl, timeout);
        this.did = did;
        this.keyPair = keyPair;
        this.platformDomain = platformDomain != null ? platformDomain
                : (did != null && !did.isEmpty() ? DomainUtils.domainFromDid(did) : "localhost");

        if (keyPair != null && did != null && !did.isEmpty()) {
            this.tokenManager = new TokenManager(
                    httpClient, keyPair, did, clientSecret, scopes, this.platformDomain);
        } else {
            this.tokenManager = null;
        }

        this.entities = new EntityApi(this);
        this.approvals = new ApprovalApi(this);
        this.revocations = new RevocationApi(this);
        this.didcomm = new DIDCommApi(this);
        this.credentials = new CredentialApi(this);
        this.discovery = new DiscoveryApi(this);
    }

    /**
     * Create a new builder for ATAPClient.
     */
    public static ATAPClientBuilder builder() {
        return new ATAPClientBuilder();
    }

    /**
     * Get the token manager for manual token operations.
     *
     * @throws IllegalStateException if authentication is not configured
     */
    public TokenManager getTokenManager() {
        if (tokenManager == null) {
            throw new IllegalStateException("Token manager not initialized. Provide did and privateKey.");
        }
        return tokenManager;
    }

    /**
     * Make an authenticated request using the token manager.
     */
    public Map<String, Object> authedRequest(String method, String path,
                                              Map<String, Object> jsonBody,
                                              byte[] rawBody,
                                              String contentType,
                                              Map<String, String> params) {
        if (tokenManager == null || keyPair == null) {
            throw new IllegalStateException(
                    "Authentication not configured. Provide did, privateKey, and optionally clientSecret.");
        }

        String accessToken = tokenManager.getAccessToken();
        return httpClient.authenticatedRequest(method, path, keyPair, accessToken,
                platformDomain, jsonBody, rawBody, contentType, params);
    }

    /**
     * Get the underlying HTTP client.
     */
    public ATAPHttpClient getHttpClient() {
        return httpClient;
    }

    /**
     * Get the DID for this client.
     */
    public String getDid() {
        return did;
    }

    /**
     * Get the platform domain.
     */
    public String getPlatformDomain() {
        return platformDomain;
    }

    @Override
    public void close() {
        // HttpClient in JDK 11 doesn't need explicit close
    }
}

File: sdks/php/src/ATAPClient.php
<?php

declare(strict_types=1);

namespace Atap\Sdk;

use Atap\Sdk\Api\ApprovalApi;
use Atap\Sdk\Api\CredentialApi;
use Atap\Sdk\Api\DIDCommApi;
use Atap\Sdk\Api\DiscoveryApi;
use Atap\Sdk\Api\EntityApi;
use Atap\Sdk\Api\RevocationApi;
use Atap\Sdk\Crypto\Ed25519;
use Atap\Sdk\Http\HttpClient;
use Atap\Sdk\OAuth\TokenManager;
use GuzzleHttp\HandlerStack;

/**
 * High-level client for the ATAP platform.
 *
 * Usage for agent/machine (client_credentials):
 *     $client = new ATAPClient([
 *         'base_url' => 'http://localhost:8080',
 *         'did' => 'did:web:localhost%3A8080:agent:abc',
 *         'private_key' => '<base64 Ed25519 seed>',
 *         'client_secret' => 'atap_...',
 *     ]);
 *
 * Usage for human/org (authorization_code + PKCE):
 *     $client = new ATAPClient([
 *         'base_url' => 'http://localhost:8080',
 *         'did' => 'did:web:localhost%3A8080:human:abc',
 *         'private_key' => '<base64 Ed25519 seed>',
 *     ]);
 *     $client->tokenManager()->obtainAuthorizationCode();
 */
class ATAPClient
{
    private HttpClient $http;
    private string $did;
    private string $platformDomain;
    private ?string $signingKey;
    private ?TokenManager $tokenManager;

    private EntityApi $entities;
    private ApprovalApi $approvals;
    private RevocationApi $revocations;
    private DIDCommApi $didcomm;
    private CredentialApi $credentials;
    private DiscoveryApi $discovery;

    /**
     * @param array<string, mixed> $config {
     *     base_url?: string,
     *     did?: string,
     *     private_key?: string,
     *     signing_key?: string (raw 64-byte secret key),
     *     client_secret?: string,
     *     scopes?: string[],
     *     platform_domain?: string,
     *     timeout?: float,
     *     handler?: HandlerStack (for testing),
     * }
     */
    public function __construct(array $config = [])
    {
        $baseUrl = (string) ($config['base_url'] ?? 'http://localhost:8080');
        $timeout = (float) ($config['timeout'] ?? 30.0);
        $handler = $config['handler'] ?? null;

        $this->http = new HttpClient($baseUrl, $timeout, $handler);
        $this->did = (string) ($config['did'] ?? '');
        $this->platformDomain = (string) ($config['platform_domain'] ?? '');

        if ($this->platformDomain === '' && $this->did !== '') {
            $this->platformDomain = DomainUtils::domainFromDid($this->did);
        }
        if ($this->platformDomain === '') {
            $this->platformDomain = 'localhost';
        }

        // Load signing key
        $this->signingKey = null;
        if (isset($config['signing_key'])) {
            $this->signingKey = (string) $config['signing_key'];
        } elseif (isset($config['private_key'])) {
            $this->signingKey = Ed25519::loadSigningKey((string) $config['private_key']);
        }

        // Initialize token manager
        $this->tokenManager = null;
        if ($this->signingKey !== null && $this->did !== '') {
            $this->tokenManager = new TokenManager(
                httpClient: $this->http,
                signingKey: $this->signingKey,
                did: $this->did,
                clientSecret: isset($config['client_secret']) ? (string) $config['client_secret'] : null,
                scopes: $config['scopes'] ?? null,
                platformDomain: $this->platformDomain,
            );
        }

        // API modules
        $this->entities = new EntityApi($this);
        $this->approvals = new ApprovalApi($this);
        $this->revocations = new RevocationApi($this);
        $this->didcomm = new DIDCommApi($this);
        $this->credentials = new CredentialApi($this);
        $this->discovery = new DiscoveryApi($this);
    }

    public function entities(): EntityApi
    {
        return $this->entities;
    }

    public function approvals(): ApprovalApi
    {
        return $this->approvals;
    }

    public function revocations(): RevocationApi
    {
        return $this->revocations;
    }

    public function didcomm(): DIDCommApi
    {
        return $this->didcomm;
    }

    public function credentials(): CredentialApi
    {
        return $this->credentials;
    }

    public function discovery(): DiscoveryApi
    {
        return $this->discovery;
    }

    /**
     * Access the token manager for manual token operations.
     */
    public function tokenManager(): TokenManager
    {
        if ($this->tokenManager === null) {
            throw new \RuntimeException('Token manager not initialized. Provide did and private_key.');
        }

        return $this->tokenManager;
    }

    /**
     * Get the HTTP client (used by API modules).
     */
    public function getHttpClient(): HttpClient
    {
        return $this->http;
    }

    /**
     * Make an authenticated request using the token manager.
     *
     * @param array<string, mixed>|null $jsonBody
     * @param array<string, string>|null $params
     * @return array<string, mixed>
     */
    public function authedRequest(
        string $method,
        string $path,
        ?array $jsonBody = null,
        ?string $rawBody = null,
        ?string $contentType = null,
        ?array $params = null,
    ): array {
        if ($this->tokenManager === null || $this->signingKey === null) {
            throw new \RuntimeException(
                'Authentication not configured. Provide did, private_key, and optionally client_secret.'
            );
        }

        $accessToken = $this->tokenManager->getAccessToken();

        return $this->http->authenticatedRequest(
            $method,
            $path,
            signingKey: $this->signingKey,
            accessToken: $accessToken,
            platformDomain: $this->platformDomain,
            jsonBody: $jsonBody,
            rawBody: $rawBody,
            contentType: $contentType,
            params: $params,
        );
    }
}
